File: app/schemas/strategy_auto_buy_scheduler.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


StrategyAutoBuyPromotionStatus = Literal[
    "pending",
    "acknowledged",
    "dismissed",
    "expired",
    "converted_to_live_attempt",
    "live_order_created",
    "live_order_synced",
    "live_order_rejected",
    "live_order_filled",
    "conversion_blocked",
    "blocked",
    "stale",
]


class StrategyAutoBuySchedulerRunRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    provider: str = "kis"
    market: str = "KR"
    symbol: str | None = None
    trigger_source: str = Field(default="manual_scheduler_dry_run", max_length=80)
    scheduler_slot: str | None = Field(default=None, max_length=80)

    @field_validator("provider")
    @classmethod
    def normalize_provider(cls, value: str) -> str:
        provider = str(value or "").strip().lower()
        if provider != "kis":
            raise ValueError("strategy auto-buy scheduler supports provider=kis only.")
        return provider

    @field_validator("market")
    @classmethod
    def normalize_market(cls, value: str) -> str:
        market = str(value or "").strip().upper()
        if market != "KR":
            raise ValueError("strategy auto-buy scheduler supports market=KR only.")
        return market

    @field_validator("symbol")
    @classmethod
    def normalize_symbol(cls, value: str | None) -> str | None:
        if value is None:
            return None
        symbol = str(value).strip().upper()
        if not symbol:
            return None
        if not symbol.isdigit() or len(symbol) > 6:
            raise ValueError("KIS symbol must be numeric.")
        return symbol.zfill(6)


class StrategyAutoBuyPromotionItem(BaseModel):
    id: int
    provider: str
    market: str
    active_profile: str | None = None
    symbol: str | None = None
    symbol_name: str | None = None
    status: str
    promotion_reason: str | None = None
    source_dry_run_signal_id: int | None = None
    source_dry_run_trade_run_id: int | None = None
    source_dry_run_order_id: int | None = None
    dry_run_action: str | None = None
    buy_score: float | None = None
    sell_score: float | None = None
    final_score: float | None = None
    confidence: float | None = None
    recommended_notional_krw: float | None = None
    simulated_quantity: float | None = None
    simulated_price: float | None = None
    simulated_notional_krw: float | None = None
    target_risk_result: dict[str, Any] = Field(default_factory=dict)
    block_reason: str | None = None
    risk_flags: list[str] = Field(default_factory=list)
    gating_notes: list[str] = Field(default_factory=list)
    expires_at: str | None = None
    acknowledged_at: str | None = None
    dismissed_at: str | None = None
    promoted_to_live_attempt_id: int | None = None
    related_live_order_id: int | None = None
    converted_live_attempt_id: int | None = None
    converted_order_id: int | None = None
    converted_at: str | None = None
    conversion_status: str | None = None
    last_sync_at: str | None = None
    last_sync_status: str | None = None
    trace_payload: dict[str, Any] = Field(default_factory=dict)
    request_payload: dict[str, Any] = Field(default_factory=dict)
    response_payload: dict[str, Any] = Field(default_factory=dict)
    created_at: str | None = None
    updated_at: str | None = None


class StrategyAutoBuyPromotionsResponse(BaseModel):
    provider: str = "kis"
    market: str = "KR"
    count: int
    items: list[StrategyAutoBuyPromotionItem] = Field(default_factory=list)
    safety: dict[str, Any] = Field(default_factory=dict)


class StrategyAutoBuyPromotionActionResponse(BaseModel):
    status: str
    promotion: StrategyAutoBuyPromotionItem
    safety: dict[str, Any] = Field(default_factory=dict)


class StrategyAutoBuyPromotionMarkConvertedRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    promoted_to_live_attempt_id: int | None = None
    related_live_order_id: int | None = None


class StrategyAutoBuySchedulerStatusResponse(BaseModel):
    provider: str = "kis"
    market: str = "KR"
    enabled: bool
    dry_run_only: bool
    promotion_queue_only: bool
    allow_live_orders: bool
    real_order_submit_allowed: bool
    active_profile: str | None = None
    allowed_profiles: list[str] = Field(default_factory=list)
    runs_today: int = 0
    max_runs_per_day: int = 0
    next_allowed_run_at: str | None = None
    min_minutes_between_runs: int = 0
    market_open: bool | None = None
    after_no_new_entry_time: bool = False
    primary_block_reason: str | None = None
    pending_promotion_count: int = 0
    latest_scheduler_run: dict[str, Any] | None = None
    schedule_slots: list[str] = Field(default_factory=list)
    safety: dict[str, Any] = Field(default_factory=dict)


class StrategyAutoBuySchedulerRunResponse(BaseModel):
    status: str
    action: str
    provider: str = "kis"
    market: str = "KR"
    active_profile: str | None = None
    dry_run_result: dict[str, Any] | None = None
    promotion: StrategyAutoBuyPromotionItem | None = None
    created_promotion: bool = False
    block_reason: str | None = None
    scheduler_run_id: int | None = None
    real_order_submitted: bool = False
    validation_called: bool = False
    broker_submit_called: bool = False
    manual_submit_called: bool = False
    safety: dict[str, Any] = Field(default_factory=dict)

File: lib/models/strategy_auto_buy_scheduler.dart
class StrategyAutoBuySchedulerStatus {
  const StrategyAutoBuySchedulerStatus({
    required this.provider,
    required this.market,
    required this.enabled,
    required this.dryRunOnly,
    required this.promotionQueueOnly,
    required this.allowLiveOrders,
    required this.realOrderSubmitAllowed,
    required this.allowedProfiles,
    required this.runsToday,
    required this.maxRunsPerDay,
    required this.minMinutesBetweenRuns,
    required this.afterNoNewEntryTime,
    required this.pendingPromotionCount,
    required this.scheduleSlots,
    required this.safety,
    this.activeProfile,
    this.nextAllowedRunAt,
    this.marketOpen,
    this.primaryBlockReason,
    this.latestSchedulerRun,
  });

  final String provider;
  final String market;
  final bool enabled;
  final bool dryRunOnly;
  final bool promotionQueueOnly;
  final bool allowLiveOrders;
  final bool realOrderSubmitAllowed;
  final String? activeProfile;
  final List<String> allowedProfiles;
  final int runsToday;
  final int maxRunsPerDay;
  final DateTime? nextAllowedRunAt;
  final int minMinutesBetweenRuns;
  final bool? marketOpen;
  final bool afterNoNewEntryTime;
  final String? primaryBlockReason;
  final int pendingPromotionCount;
  final Map<String, dynamic>? latestSchedulerRun;
  final List<String> scheduleSlots;
  final Map<String, dynamic> safety;

  factory StrategyAutoBuySchedulerStatus.fromJson(Map<String, dynamic> json) {
    return StrategyAutoBuySchedulerStatus(
      provider: _string(json['provider'], 'kis'),
      market: _string(json['market'], 'KR'),
      enabled: json['enabled'] == true,
      dryRunOnly: json['dry_run_only'] != false,
      promotionQueueOnly: json['promotion_queue_only'] != false,
      allowLiveOrders: json['allow_live_orders'] == true,
      realOrderSubmitAllowed: json['real_order_submit_allowed'] == true,
      activeProfile: _nullableString(json['active_profile']),
      allowedProfiles: _strings(json['allowed_profiles']),
      runsToday: _int(json['runs_today']),
      maxRunsPerDay: _int(json['max_runs_per_day']),
      nextAllowedRunAt: _dateTime(json['next_allowed_run_at']),
      minMinutesBetweenRuns: _int(json['min_minutes_between_runs']),
      marketOpen: _nullableBool(json['market_open']),
      afterNoNewEntryTime: json['after_no_new_entry_time'] == true,
      primaryBlockReason: _nullableString(json['primary_block_reason']),
      pendingPromotionCount: _int(json['pending_promotion_count']),
      latestSchedulerRun: _nullableMap(json['latest_scheduler_run']),
      scheduleSlots: _strings(json['schedule_slots']),
      safety: _map(json['safety']),
    );
  }

  StrategyAutoBuySchedulerStatus copyWith({
    String? provider,
    String? market,
    bool? enabled,
    bool? dryRunOnly,
    bool? promotionQueueOnly,
    bool? allowLiveOrders,
    bool? realOrderSubmitAllowed,
    String? activeProfile,
    List<String>? allowedProfiles,
    int? runsToday,
    int? maxRunsPerDay,
    DateTime? nextAllowedRunAt,
    int? minMinutesBetweenRuns,
    bool? marketOpen,
    bool? afterNoNewEntryTime,
    String? primaryBlockReason,
    bool clearPrimaryBlockReason = false,
    int? pendingPromotionCount,
    Map<String, dynamic>? latestSchedulerRun,
    List<String>? scheduleSlots,
    Map<String, dynamic>? safety,
  }) {
    return StrategyAutoBuySchedulerStatus(
      provider: provider ?? this.provider,
      market: market ?? this.market,
      enabled: enabled ?? this.enabled,
      dryRunOnly: dryRunOnly ?? this.dryRunOnly,
      promotionQueueOnly: promotionQueueOnly ?? this.promotionQueueOnly,
      allowLiveOrders: allowLiveOrders ?? this.allowLiveOrders,
      realOrderSubmitAllowed:
          realOrderSubmitAllowed ?? this.realOrderSubmitAllowed,
      activeProfile: activeProfile ?? this.activeProfile,
      allowedProfiles: allowedProfiles ?? this.allowedProfiles,
      runsToday: runsToday ?? this.runsToday,
      maxRunsPerDay: maxRunsPerDay ?? this.maxRunsPerDay,
      nextAllowedRunAt: nextAllowedRunAt ?? this.nextAllowedRunAt,
      minMinutesBetweenRuns:
          minMinutesBetweenRuns ?? this.minMinutesBetweenRuns,
      marketOpen: marketOpen ?? this.marketOpen,
      afterNoNewEntryTime: afterNoNewEntryTime ?? this.afterNoNewEntryTime,
      primaryBlockReason: clearPrimaryBlockReason
          ? null
          : primaryBlockReason ?? this.primaryBlockReason,
      pendingPromotionCount:
          pendingPromotionCount ?? this.pendingPromotionCount,
      latestSchedulerRun: latestSchedulerRun ?? this.latestSchedulerRun,
      scheduleSlots: scheduleSlots ?? this.scheduleSlots,
      safety: safety ?? this.safety,
    );
  }
}

class StrategyAutoBuySchedulerRunResult {
  const StrategyAutoBuySchedulerRunResult({
    required this.status,
    required this.action,
    required this.provider,
    required this.market,
    required this.createdPromotion,
    required this.realOrderSubmitted,
    required this.validationCalled,
    required this.brokerSubmitCalled,
    required this.manualSubmitCalled,
    required this.safety,
    this.activeProfile,
    this.dryRunResult,
    this.promotion,
    this.blockReason,
    this.schedulerRunId,
  });

  final String status;
  final String action;
  final String provider;
  final String market;
  final String? activeProfile;
  final Map<String, dynamic>? dryRunResult;
  final Map<String, dynamic>? promotion;
  final bool createdPromotion;
  final String? blockReason;
  final int? schedulerRunId;
  final bool realOrderSubmitted;
  final bool validationCalled;
  final bool brokerSubmitCalled;
  final bool manualSubmitCalled;
  final Map<String, dynamic> safety;

  factory StrategyAutoBuySchedulerRunResult.fromJson(
      Map<String, dynamic> json) {
    return StrategyAutoBuySchedulerRunResult(
      status: _string(json['status'], 'unknown'),
      action: _string(json['action'], 'blocked'),
      provider: _string(json['provider'], 'kis'),
      market: _string(json['market'], 'KR'),
      activeProfile: _nullableString(json['active_profile']),
      dryRunResult: _nullableMap(json['dry_run_result']),
      promotion: _nullableMap(json['promotion']),
      createdPromotion: json['created_promotion'] == true,
      blockReason: _nullableString(json['block_reason']),
      schedulerRunId: _nullableInt(json['scheduler_run_id']),
      realOrderSubmitted: json['real_order_submitted'] == true,
      validationCalled: json['validation_called'] == true,
      brokerSubmitCalled: json['broker_submit_called'] == true,
      manualSubmitCalled: json['manual_submit_called'] == true,
      safety: _map(json['safety']),
    );
  }
}

String _string(Object? value, String fallback) {
  final text = value?.toString().trim();
  return text == null || text.isEmpty ? fallback : text;
}

String? _nullableString(Object? value) {
  final text = value?.toString().trim();
  return text == null || text.isEmpty || text == 'null' ? null : text;
}

bool? _nullableBool(Object? value) {
  if (value is bool) return value;
  return null;
}

int _int(Object? value) => _nullableInt(value) ?? 0;

int? _nullableInt(Object? value) {
  if (value == null) return null;
  if (value is num) return value.toInt();
  return int.tryParse(value.toString().trim());
}

DateTime? _dateTime(Object? value) {
  final text = value?.toString().trim();
  if (text == null || text.isEmpty || text == 'null') return null;
  return DateTime.tryParse(text);
}

Map<String, dynamic> _map(Object? value) {
  return value is Map ? Map<String, dynamic>.from(value) : const {};
}

Map<String, dynamic>? _nullableMap(Object? value) {
  return value is Map ? Map<String, dynamic>.from(value) : null;
}

List<String> _strings(Object? value) {
  if (value is! List) return const [];
  return [for (final item in value) item.toString()];
}
